app/src/main/java/io/appium/uiautomator2/utils/ElementHelpers.java

```diff
@@ -169,13 +171,10 @@ public static void setText(final Object element, final String text, final boolea
             textToSend = UnicodeEncoder.encode(textToSend);
             Logger.debug("Encoded text: " + textToSend);
         }
+
         Logger.debug("Sending text to element: " + textToSend);
-        if (element instanceof UiObject2) {
-            UiObject2.class.cast(element).setText(textToSend);
-        } else if (element instanceof UiObject) {
-            UiObject.class.cast(element).setText(textToSend);
-        } else {
-            throw new UiAutomator2Exception("Unknown object type: " + element.getClass().getName());
-        }
+        Bundle args = new Bundle();
+        args.putCharSequence(AccessibilityNodeInfo.ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE, textToSend);
+        return nodeInfo.performAction(AccessibilityNodeInfo.ACTION_SET_TEXT, args);
```

Comment:
interesting, so we're switching to `performAction` rather than using the uiobject methods. I'm OK with this as long as the behavior is the same as before!

Reply:
this code is called for both cases if you check the downstream implantation if OS API version is above KitKat

Reply:
It's ok for ```UiObject2```. But for ```UiObject``` we will get a problem here:
1. ```findAccessibilityNodeInfo``` returns null if element does not exist
2. ```setText``` waits for element with ```Configurator.getWaitForSelectorTimeout()``` (I'am ok if we skip this part)

https://github.com/appium/appium-uiautomator2-server/blob/665f99151b2ef2f658d7555c40493c7c5dfb659a/app/src/main/java/io/appium/uiautomator2/core/AccessibilityNodeInfoGetter.java#L33

https://android.googlesource.com/platform/frameworks/uiautomator/+/android-support-test/src/main/java/android/support/test/uiautomator/UiObject.java#184

Reply:
added null verification